### PR TITLE
Copy execute permissions for apphost for non-Windows

### PR DESCRIPTION
**Pull request by steveharter**
@dotnet/dotnet-cli

@MattGertz for approval

**Customer scenario**
The executable generated from `dotnet publish -r{RID}` doesn't have the "x" (execute) permission for non-Windows and thus when attempting to execute, a `Permission denied` error is returned from the OS. This means the standalone + RID scenario is broken for non-Windows.

The fix is to do a File.Copy from the source apphost to the destination (modified apphost) so the execute permission will be copied with that. Then the file is updated to have the proper contents.

**Bugs this fixes:**
https://github.com/dotnet/sdk/issues/1331

**Workarounds, if any**
The developer would need to do a manual chmod, such as `chmod +x myConsoleApp`

**Risk**
Low.

**Performance impact**
Minor; extra copy of ~77K apphost file. Created issue https://github.com/dotnet/sdk/issues/1339 to address in future.

**Is this a regression from a previous update?**
No

**Root cause analysis:**
The underlying issue has been masked for a long time by a NuGet issue which didn't restore executable files with the "x" permission.

**How was the bug found?**
Going through a core developer scenario on non-Windows.

### CODE REVIEW

File: src/Tasks/Microsoft.NET.Build.Tasks/EmbedAppNameInHost.cs
Comment (gkhanna79) [inline src/Tasks/Microsoft.NET.Build.Tasks/EmbedAppNameInHost.cs]: Why do we need this "if" check. The "else" section would work for Windows as well and keep the implementation consistent.

Comment (steveharter) [inline src/Tasks/Microsoft.NET.Build.Tasks/EmbedAppNameInHost.cs]: The 'if' should not be necessary, although keeping it will prevent the (unnecessary) file copy for Windows.

Comment (gkhanna79) [inline src/Tasks/Microsoft.NET.Build.Tasks/EmbedAppNameInHost.cs]: My suggestion would be to keep the code consistent - there is no overhead scenario in this operation.

**Comment by nguerrera** [comment]
Rebase and retarget PR to release/2.0

File: src/Tasks/Microsoft.NET.Build.Tasks/EmbedAppNameInHost.cs
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/EmbedAppNameInHost.cs]: I would keep the code consistent between platforms too. I don't know what you mean by "no overhead", though. We're copying the whole file only to replace its contents completely. We can fix this by finding the index of the marker, seeking to the right position and overwriting it.

On a related note, I pointed out in the original code review that reading the whole file in to a managed byte array was wasteful, but that was never addressed.

**Review by nguerrera — CHANGES_REQUESTED** [review]
See above

File: src/Tasks/Microsoft.NET.Build.Tasks/EmbedAppNameInHost.cs
Comment (steveharter) [inline src/Tasks/Microsoft.NET.Build.Tasks/EmbedAppNameInHost.cs]: OK I will remove the 'if', and log a future issue for the performance aspect so we only change the marker and not the whole file. The apphost file is ~77K so fairly small, but still we should do the right thing here.

Comment (gkhanna79) [inline src/Tasks/Microsoft.NET.Build.Tasks/EmbedAppNameInHost.cs]: Thanks @steveharter - I concur with you.

Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/EmbedAppNameInHost.cs]: Sounds good. Thanks.

**Review by gkhanna79 — APPROVED** [review]

**Comment by steveharter** [comment]
>> Rebase and retarget PR to release/2.0

@nguerrera I was assuming we want to go into master first, then release/2.0. If not, are we planning on doing an RI later?

**Comment by nguerrera** [comment]
> @nguerrera I was assuming we want to go into master first, then release/2.0. If not, are we planning on doing an RI later?

We will RI 2.0 -> master (work is in progress to make that automatic and often). We should avoid dual checkins to master and release where possible.

**Comment by nguerrera** [comment]
This needs to be rebased to just your changes over 2.0. As is, it is merging changes for 2.1 from master.

**Comment by steveharter** [comment]
Yeah I think I’m good now. Thanks

**Review by nguerrera — APPROVED** [review]

**Comment by nguerrera** [comment]
Please fill update the PR description with the template as seen in https://github.com/dotnet/sdk/pull/1326#issue-234692463

Once we have that, we can tag Matt for approval and merge if approved.

**Comment by nguerrera** [comment]
Also, is the perf bug for future filed?

**Comment by steveharter** [comment]
@nguerrera just created perf issue https://github.com/dotnet/sdk/issues/1339

**Review by livarcocc — APPROVED** [review]

**Comment by steveharter** [comment]
@nguerrera should I also send out a shiproom mail?

**Comment by nguerrera** [comment]
@steveharter No, this is sufficient for a change affecting only 2.0 SDK.

**Comment by nguerrera** [comment]
@MattGertz for approval

**Comment by gkhanna79** [comment]
CC @leecow as well to keep this in his purview as well.

**Pull request merged**
